pubsub/awssnssqs/awssnssqs.go

```diff
@@ -452,46 +457,78 @@ func openSQSTopic(ctx context.Context, sess client.ConfigProvider, qURL string,
 
 // SendBatch implements driver.Topic.SendBatch.
 func (t *sqsTopic) SendBatch(ctx context.Context, dms []*driver.Message) error {
-	if len(dms) != 1 {
-		panic("sqsTopic.SendBatch should only get one message at a time")
-	}
-	dm := dms[0]
-	attrs := map[string]*sqs.MessageAttributeValue{}
-	for k, v := range encodeMetadata(dm.Metadata) {
-		attrs[k] = &sqs.MessageAttributeValue{
-			DataType:    stringDataType,
-			StringValue: aws.String(v),
-		}
-	}
-	body, didEncode := maybeEncodeBody(dm.Body, t.opts.BodyBase64Encoding)
-	if didEncode {
-		attrs[base64EncodedKey] = &sqs.MessageAttributeValue{
-			DataType:    stringDataType,
-			StringValue: aws.String("true"),
+	req := &sqs.SendMessageBatchInput{
+		QueueUrl: aws.String(t.qURL),
+	}
+	for _, dm := range dms {
+		attrs := map[string]*sqs.MessageAttributeValue{}
+		for k, v := range encodeMetadata(dm.Metadata) {
+			attrs[k] = &sqs.MessageAttributeValue{
+				DataType:    stringDataType,
+				StringValue: aws.String(v),
+			}
 		}
-	}
-	if len(attrs) == 0 {
-		attrs = nil
-	}
-	req := &sqs.SendMessageInput{
-		QueueUrl:          aws.String(t.qURL),
-		MessageAttributes: attrs,
-		MessageBody:       aws.String(body),
-	}
-	if dm.BeforeSend != nil {
-		asFunc := func(i interface{}) bool {
-			if p, ok := i.(**sqs.SendMessageInput); ok {
-				*p = req
-				return true
+		body, didEncode := maybeEncodeBody(dm.Body, t.opts.BodyBase64Encoding)
+		if didEncode {
+			attrs[base64EncodedKey] = &sqs.MessageAttributeValue{
+				DataType:    stringDataType,
+				StringValue: aws.String("true"),
 			}
-			return false
 		}
-		if err := dm.BeforeSend(asFunc); err != nil {
-			return err
+		if len(attrs) == 0 {
+			attrs = nil
+		}
+		entry := &sqs.SendMessageBatchRequestEntry{
+			Id:                aws.String(strconv.Itoa(len(req.Entries))),
+			MessageAttributes: attrs,
+			MessageBody:       aws.String(body),
+		}
+		req.Entries = append(req.Entries, entry)
+		if dm.BeforeSend != nil {
+			// A previous revision used the non-batch API SendMessage, which takes
+			// a *sqs.SendMessageInput. For backwards compatibility for As, continue
+			// to support that type. If it is requested, create a SendMessageInput
+			// with the fields from SendMessageBatchRequestEntry that were set, and
+			// then copy all of the matching fields back after calling dm.BeforeSend.
+			var smi *sqs.SendMessageInput
+			asFunc := func(i interface{}) bool {
+				if p, ok := i.(**sqs.SendMessageInput); ok {
+					smi = &sqs.SendMessageInput{
+						// Id does not exist on SendMessageInput.
+						MessageAttributes: entry.MessageAttributes,
+						MessageBody:       entry.MessageBody,
+					}
+					*p = smi
+					return true
+				}
+				if p, ok := i.(**sqs.SendMessageBatchRequestEntry); ok {
+					*p = entry
+					return true
+				}
+				return false
+			}
+			if err := dm.BeforeSend(asFunc); err != nil {
+				return err
+			}
+			if smi != nil {
+				// Copy all of the fields that may have been modified back to the entry.
```

Comment:
Add a comment to mention this is for backward-compatibility and the two types between `entry` and `smi` have matching fields so that you can copy them back instead of sending those single inputs one by one.

Reply:
I already have that comment, up on line 488. I don't think it's useful to copy it here...?

I am not sure what you mean by "instead of sending those single inputs one by one".

Reply:
I mean if they are not exactly matching with each other, you would need to call the single RPC's on `smi`s for the user calling `as` on the old type. I think the comment above could be sufficient.